### ISSUE TITLE
Request callback documentation confusing or incorrect

### ISSUE THREAD

**Issue by jcoffland** [libevent/libevent#589]
The documentation for ``evhttp_request_set_error_cb()`` says:
```
On error, both the error callback and the regular callback will be called, error callback is called before the regular callback. 
```

It is unclear what is meant by the "regular callback".  I had assumed this meant the request complete callback.  However, it's not possible for any request callback to be called after the error callback.  The error callback is invoked in ``evhttp_connection_fail_()`` and the request is freed in that function.  So it is not possible to call the complete callback or any other request callback after the error callback is called.

The documentation for ``evhttp_request_set_on_complete_cb()`` does not make it clear that the complete callback may never get called.  This is problematic if the complete callback is being used to deallocate resources.

Further complicating the resource allocation issue, if the user cancels the request with ``evhttp_cancel_request()`` then neither callback is called.  A much cleaner solution is to add a request free callback which is always called in ``evhttp_request_free()`` if set by the user.

**Comment by azat** [comment]
Indeed it has a lot of confusion. But by "regular callback" the callback passed with `evhttp_request_new()` was meant.

It would be great if you will be interesting in improving this, any chances?

**Comment by ploxiln** [comment]
I noticed that the "regular callback" is called, with `NULL` for the request struct pointer, if the request times-out. It seems to be called with the real request struct pointer, but with a response_code of 0, if the connection immediately failed. I'm using libevent-2.1.8

**Comment by azat** [comment]
Yeah, there was some weirdness like this, AFAIR. And agree, different callbacks is really uncomfortable for use. since you have to introduce your own context to store errors and other stuff.

**Comment by jcoffland** [comment]
I could submit a PR.  I'm currently working with an older version.  I need to first update my code base.

**Comment by azat** [comment]
@jcoffland any news?

**Comment by jcoffland** [comment]
Producing the documentation itself is not a problem.  I haven't had the time to figure out your documentation system.

**Comment by azat** [comment]
It is just comments from the [code](https://github.com/libevent/libevent/blob/69bc2da79fb2f358c22feba45e9ad4752729d320/include/event2/http.h#L625), the [documentation](http://www.wangafu.net/~nickm/libevent-2.1/doxygen/html/http_8h.html#a90bf87f4fdaea3ccf414bf8572b0228a) will be automatically regenerated via doxygen

**Comment by jcoffland** [comment]
I did submit https://github.com/libevent/libevent/pull/591and it has documentation.  If that PR is accepted then this issue is resolved as far as I'm concerned.
